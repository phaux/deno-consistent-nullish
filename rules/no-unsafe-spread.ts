import { trimRange } from "../util/trimRange.ts";

export const noUnsafeSpreadRule: Deno.lint.Rule = {
  create: (ctx) => ({
    ObjectExpression(objNode) {
      for (const [propIdx, propNode] of objNode.properties.entries()) {
        if (propIdx <= 0) continue;
        if (propNode.type !== "SpreadElement") continue;

        ctx.report({
          range: trimRange(ctx.sourceCode, [
            propNode.range[0],
            propNode.argument.range[0],
          ]),
          message: "Object spread is only allowed as the first element.",
          hint:
            "Object spread is broken in TypeScript without `exactOptionalPropertyTypes` mode. " +
            "Please assign each property explicitly.",
        });
      }
    },
  }),
};
