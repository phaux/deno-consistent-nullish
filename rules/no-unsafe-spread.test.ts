import { assertEquals } from "@std/assert";
import { assertSnapshot } from "@std/testing/snapshot";
import { noUnsafeSpreadRule } from "./no-unsafe-spread.ts";

const plugin: Deno.lint.Plugin = {
  name: "test",
  rules: { rule: noUnsafeSpreadRule },
};

Deno.test("bans spread in objects", async (t) => {
  const diagnostics = Deno.lint.runPlugin(
    plugin,
    "test.ts",
    `
    ({ a: 1, ...obj });
    ({ a: 1, b: 2, ...obj });
    ({ a: 1, ...obj, c: 3 });
    ({ obj, ...obj });
    ({ ...obj1, ...obj2, ...obj3 });
    `
  );
  assertEquals(diagnostics.length, 6);
  await assertSnapshot(t, diagnostics);
});

Deno.test("allows spread in non-object", () => {
  const diagnostics = Deno.lint.runPlugin(
    plugin,
    "test.ts",
    `
    ([1, ...arr]);
    f(1, 2, ...arr);
    ([1, ...arr, 3]);
    f(arr, ...arr);
    ([...arr1, ...arr2]);
    `
  );
  assertEquals(diagnostics.length, 0);
});

Deno.test("allows spread as first element", () => {
  const diagnostics = Deno.lint.runPlugin(
    plugin,
    "test.ts",
    `
    ({ ...obj, b: 2 });
    ({ ...obj, b: 2, c: 3 });
    ({ ...obj });
    `
  );
  assertEquals(diagnostics.length, 0);
});
