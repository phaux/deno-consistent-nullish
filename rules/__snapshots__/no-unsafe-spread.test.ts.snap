export const snapshot = {};

snapshot[`bans spread in objects 1`] = `
[
  {
    fix: [],
    hint: "Object spread is broken in TypeScript without \`exactOptionalPropertyTypes\` mode. You should assign each property explicitly.",
    id: "test/rule",
    message: "Object spread is only allowed as the first element.",
    range: [
      14,
      17,
    ],
  },
  {
    fix: [],
    hint: "Object spread is broken in TypeScript without \`exactOptionalPropertyTypes\` mode. You should assign each property explicitly.",
    id: "test/rule",
    message: "Object spread is only allowed as the first element.",
    range: [
      44,
      47,
    ],
  },
  {
    fix: [],
    hint: "Object spread is broken in TypeScript without \`exactOptionalPropertyTypes\` mode. You should assign each property explicitly.",
    id: "test/rule",
    message: "Object spread is only allowed as the first element.",
    range: [
      68,
      71,
    ],
  },
  {
    fix: [],
    hint: "Object spread is broken in TypeScript without \`exactOptionalPropertyTypes\` mode. You should assign each property explicitly.",
    id: "test/rule",
    message: "Object spread is only allowed as the first element.",
    range: [
      97,
      100,
    ],
  },
  {
    fix: [],
    hint: "Object spread is broken in TypeScript without \`exactOptionalPropertyTypes\` mode. You should assign each property explicitly.",
    id: "test/rule",
    message: "Object spread is only allowed as the first element.",
    range: [
      124,
      127,
    ],
  },
  {
    fix: [],
    hint: "Object spread is broken in TypeScript without \`exactOptionalPropertyTypes\` mode. You should assign each property explicitly.",
    id: "test/rule",
    message: "Object spread is only allowed as the first element.",
    range: [
      133,
      136,
    ],
  },
]
`;
